POEM_003.md

```diff
@@ -0,0 +1,111 @@
+POEM ID: 003  
+Title: Adding a pre_setup method to create objects and communicators  
+Author: anilyil (Anil Yildirim)  
+Competing POEMs: N/A  
+Related POEMs: N/A  
+Associated implementation PR: N/A  
+
+Status: 
+
+ - [x] Active
+ - [ ] Requesting decision
+ - [ ] Accepted
+ - [ ] Rejected
+
+
+Problem statement
+=================
+
+OpenMDAO problems contain a method called `setup`, which is called by the user after the user specifies all components and connections.
+No fundamental changes to the components can be done after this method is called; in particular, sizes of the intputs and outputs for any component cannot be changed.
+Furthermore, the communicators are not created until `setup` is called, therefore users cannot initialize supplementary Python objects that require access to the communicator.
+
+Frequently, these objects that require the communicator needs to be initialized to determine the input and output sizes for their components, effectively creating a circular dependency.
+This factor makes it difficult to integrate complied analysis codes in OpenMDAO components in a modular way.
+This problem is currently avoided by ad hoc implementations, where the Python objects pass information in a layer hidden from OpenMDAO, which is not ideal and limits the flexibility of the framework.
+
+
+
+
+Suggested solution
+==================
+
+We suggest adding a `pre_setup` method to the `Problem` class in OpenMDAO.
+This method will:
+
+1. Create the communicators,
+2. traverse down the model hierarchy and call the `pre_setup` method on all components, and
+3. traverse up the model hierarchy and call the `pre_configure` method on all components.
+
```

Comment:
This description currently matches what is done in the `setup` method. When combined with the existing `configure` method I think you have all the functionality you need.

Reply:
Currently, we cannot add new i/o and make new connections after the p.setup() call finishes. We want to be able to create new i/o, and make new connections outside the p.setup() call.

Reply:
ok, so in the case you're presenting, the user does all this in the run-script itself, after they call problem setup. 

You could currently work around this by defining a group sub-class in your run script where you put your setup and configure actions. Then set that group as the root of the problem. That would give you the desired functionality. However, you could potentially argue that its a bit clunky that way. 

So clearly the functionality already exists in OM, but it seems like this POEM is asking for more granular access to the setup stack then, to access the needed functionality without having to sub-class anything.

Reply:
> ok, so in the case you're presenting, the user does all this in the run-script itself, after they call problem setup.
> 
> You could currently work around this by defining a group sub-class in your run script where you put your setup and configure actions. Then set that group as the root of the problem. That would give you the desired functionality. However, you could potentially argue that its a bit clunky that way.
>
> So clearly the functionality already exists in OM, but it seems like this POEM is asking for more granular access to the setup stack then, to access the needed functionality without having to sub-class anything.

Exactly. We can get around it, but it creates additional levels that are not necessary and complicates the run scripts.